test/emqx_protocol_SUITE.erl

```diff
@@ -19,114 +20,212 @@
-%%     Msg = emqx_message:from_packet(?PUBLISH_PACKET(1, <<"topic">>, 10, <<"payload">>)),
-%%     ?assertEqual(1, Msg#message.qos),
-%%     %% ?assertEqual(10, Msg#message.pktid),
-%%     ?assertEqual(<<"topic">>, Msg#message.topic),
-%%     WillMsg = emqx_message:from_packet(#mqtt_packet_connect{will_flag  = true,
-%%                                                             will_topic = <<"WillTopic">>,
-%%                                                             will_payload   = <<"WillMsg">>}),
-%%     ?assertEqual(<<"WillTopic">>, WillMsg#message.topic),
-%%     ?assertEqual(<<"WillMsg">>, WillMsg#message.payload).
-
-    %% Msg2 = emqx_message:fomat_packet(<<"username">>, <<"clientid">>,
-    %%                                   ?PUBLISH_PACKET(1, <<"topic">>, 20, <<"payload">>)),
+%% receive_messages(0, Msgs) ->
+%%     Msgs;
+%% receive_messages(Count, Msgs) ->
+%%     receive
+%%         Other ->
+%%             ct:log("~p~n", [Other]),
+%%             receive_messages(Count-1, Msgs)
+%%     after 10 ->
+%%         Msgs
+%%     end.
 
+all() ->
+    [
+     {group, mqttv4},
+     {group, mqttv5}].
 
+groups() ->
+    [{mqttv4,
+      [sequence],
+      [
+       connect_v4,
+       subscribe_v4
+       %% unsubscribe_v4
+       %% publish_v4-
+      ]},
+     {mqttv5,
+      [sequence],
+      [
+       connect_v5,
+       subscribe_v5
+      ]
+     }].
+
+init_per_suite(Config) ->
+    emqx_ct_broker_helpers:run_setup_steps(),
+    Config.
+
+end_per_suite(_Config) ->
+    emqx_ct_broker_helpers:run_teardown_steps().
+
+connect_v4(_) ->
+    ct:print("MQTT v4 connect test starting"),
+    {ok, Sock} = emqx_client_sock:connect({127,0,0,1}, 1883, [binary, {packet, raw}, {active, false}], 3000),
+    ConnectPacket = raw_send_serialize(?PACKET(?PUBLISH)),
+    emqx_client_sock:send(Sock, ConnectPacket),
+    {error, closed} = gen_tcp:recv(Sock, 0),
+    emqx_client_sock:close(Sock),
+    {ok, Sock3} = emqx_client_sock:connect({127,0,0,1}, 1883, [binary, {packet, raw}, {active, false}], 3000),
+    ConnectPacket3 = raw_send_serialize(?CONNECT_PACKET(#mqtt_packet_connect{
+                                                           client_id  = <<"mqttv4_client">>,
+                                                           username = <<"admin">>,
+                                                           password = <<"public">>,
+                                                           proto_ver  = ?MQTT_PROTO_V4})),
+    emqx_client_sock:send(Sock3, ConnectPacket3),
+    {ok, Data} = gen_tcp:recv(Sock3, 0),
+    {ok, ?CONNACK_PACKET(?CONNACK_ACCEPT), _} = raw_recv_parse(Data, ?MQTT_PROTO_V4),
+
+    emqx_client_sock:send(Sock3, ConnectPacket3),
+    {error, closed} = gen_tcp:recv(Sock3, 0),
+    emqx_client_sock:close(Sock3),
+    ok.
+
+
+connect_v5(_) ->
+    ct:print("MQTT v5 connect test starting"),
+    {ok, Sock1} = emqx_client_sock:connect({127,0,0,1}, 1883, [binary, {packet, raw}, {active, false}], 3000),
```

Comment:
tests seem to very similar.
we can write a help function like this:
```
with_connection(IP, Port, Opts, Timeout, DoFun) ->
  {ok, Sock} = emqx_client_sock:connect({127,0,0,1}, 1883, [binary, {packet, raw}, {active, false}], 3000),
  try
    DoFun(Sock)
  after
    emqx_client_sock:close(Sock)
  end.
```

Reply:
Yes, good idea.